src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -1003,212 +1107,220 @@ class FrontendGenImpl {
+    // will populate grap.value_info() with more information than we can
+    // get from InferShapeForFunctionNode(functionProto,..).
+    onnx::GraphProto graph;
+
+    for (int i = 0; i < functionProto.input_size(); ++i) {
+      onnx::ValueInfoProto *info = graph.add_input();
+      info->set_name(functionProto.input(i));
+      // Set type if known, otherwise it's uninitialized which means NoneType.
+      if (i < node.input_size())
+        *info->mutable_type() = inputOnnxTypes[i];
+    }
+
+    for (int i = 0; i < functionProto.output_size(); ++i) {
+      onnx::ValueInfoProto *info = graph.add_output();
+      info->set_name(functionProto.output(i));
+      // Set type if known, otherwise it's uninitialized which means NoneType.
+      if (i < node.output_size()) {
+        if (const onnx::TypeProto *onnxType =
+                onnx_type_map.GetByOnnxName(node.output(i))) {
+          *info->mutable_type() = *onnxType;
+        }
       }
     }
-  }
 
-  bool TryImportFunctionCallNode(const onnx::NodeProto &node) {
-    const onnx::OpSchema *schema = GetOpSchema(node);
-    if (schema == nullptr)
-      return false;
+    *graph.mutable_node() = std::move(functionProto.node());
 
-    // Collect input/output MLIR types, input ONNX types, and input MLIR values.
-    // TODO: Optional inputs/outputs of functions not handled yet.
-    onnx::TypeProto unspecifiedType;
-    llvm::SmallVector<Type, 16> operandTypes;
-    llvm::SmallVector<Type, 16> resultTypes;
-    llvm::SmallVector<::Value, 16> operands;
-    std::vector<onnx::TypeProto> operandOnnxTypes;
-
-    for (auto &v : node.input()) {
-      if (v.empty()) {
-        // Missing (optional) parameter.
-        operandOnnxTypes.push_back(unspecifiedType);
-        operands.push_back(createNoneValue());
-        operandTypes.push_back(builder_.getNoneType());
-        continue;
-      }
-      // Function translation requires input (onnx) types
-      if (const onnx::TypeProto *onnxTypePtr = onnx_type_map.GetByOnnxName(v)) {
-        operandOnnxTypes.push_back(*onnxTypePtr);
-        auto val = LookupOnnxName(v);
-        operands.push_back(val);
-        operandTypes.push_back(val.getType());
-      } else {
-        return false;
-      }
-    }
-    for (auto &v : node.output()) {
-      if (v.empty())
-        continue;
-      if (const onnx::TypeProto *onnxTypePtr = onnx_type_map.GetByOnnxName(v)) {
-        auto resultType = ImportType(*onnxTypePtr);
-        resultTypes.push_back(resultType);
-      } else {
-        return false;
-      }
+    // Substitute caller attributes in graph nodes:
+    AttrMap caller_attr_map;
+    for (const onnx::AttributeProto &attr : node.attribute())
+      caller_attr_map[attr.name()] = &attr;
+    AttrMap attr_map;
+    for (const std::string &attrName : functionProto.attribute()) {
+      auto it = caller_attr_map.find(attrName);
+      if (it != caller_attr_map.end())
+        attr_map[attrName] = it->second;
```

Comment:
I am not familiar with the attribute substitution. Any link to the relevant document?

Reply:
there isn't a lot of documentation

FunctionProto has attributes: https://github.com/onnx/onnx/blob/ad834eb73ee0cd9b6fa9ea892caeed5fa17d7dc0/onnx/onnx-ml.proto#L798-L799

and onnx's shape inference implementation illustrates how they are used:
https://github.com/onnx/onnx/blob/ad834eb73ee0cd9b6fa9ea892caeed5fa17d7dc0/onnx/shape_inference/implementation.cc#L524-L563

Reply:
`code_function_with_attributes` in FunModelTest in onnx's ParserTest is a good illustration:
https://github.com/onnx/onnx/blob/main/onnx/test/cpp/parser_test.cc#L500-L524

Reply:
It seems to me that the attributes with the call site is always used and the attributes attached with the Function is useless.

Reply:
Can the attribute of a function be used by ops inside function body? Or it is actually used for the call site? I am ponding on the comment

>  so a separate MLIR function more or less needs to be created for each call site,

Reply:
if a FunctionProto has an attribute gamma then it will be referenced as @gamma in the function's body and will be substituted by the caller's gamma attribute value when invoked

in the example below 3.0 is substituted for @gamma in foo's body in `out = custom_domain.foo<alpha=2.0, gamma=3.0>(x)` and if foo is called elsewhere with a different gamma then a different substitution needs to happen, and mlir functions don't support the substitution of values for attributes and therefore foo cannot be represented by a single mlir function

```
<
  ir_version: 9,
  opset_import: [ "" : 15, "custom_domain" : 1]
>
agraph (float[N] x) => (float[N] out)
{
  out = custom_domain.foo<alpha=2.0, gamma=3.0>(x)
}

<
domain: "custom_domain",
opset_import: [ "" : 15],
doc_string: "function foo"
>
  foo
  <alpha: float=4.0, gamma>
  (X) => (C)
  {
      constant_alpha = Constant<value_float: float=@alpha>()
      constant_gamma = Constant<value_float: float=@gamma>()
      constant_alpha_x = Mul(constant_alpha, X)
      C = Add(constant_alpha_x, constant_gamma)
  }
```

Reply:
> It seems to me that the attributes with the call site is always used and the attributes attached with the Function is useless.

the "attributes" attached with the Function are just attribute names

Reply:
Will the attribute binding also use this mapping? string to Value?
```
void BindOnnxName(const std::string &onnx_name, Value symbol) {
     frontend_symbols_.AddMapping(onnx_name, symbol);
   }
```

Reply:
no, attributes names map to constants (onnx::AttributeProto which will be translated to mlir::NamedAttribute in convertOnnxAttributeProtoToMlirNamedAttribute) not mlir::Value, and attribute `@` references (like `@gamma` in the example) are replaced by substitution in one shot in `replaceAttrRefs(graph, attr_map)`, there is no FrontendGenImpl instance member for attributes like frontend_symbols_ for values